c/src/channel.rs

```diff
@@ -410,6 +439,69 @@ unsafe fn do_foxglove_mcap_write_metadata(
     Ok(())
 }
 
+/// Write an attachment to an MCAP file.
+///
+/// Attachments are arbitrary binary data that can be stored alongside messages.
+/// Common uses include storing configuration files, calibration data, or other
+/// reference material related to the recording.
+///
+/// Returns 0 on success, or returns a FoxgloveError code on error.
+///
+/// # Safety
+/// `writer` must be a valid pointer to a `FoxgloveMcapWriter` created via `foxglove_mcap_open`.
+/// `attachment` must be a valid pointer to a `FoxgloveMcapAttachment`.
+/// The `name` and `media_type` fields of the attachment must be valid UTF-8 strings.
+/// The `data` field must be a valid pointer to an array of bytes with length `data_len`,
+/// or null if `data_len` is 0.
+#[unsafe(no_mangle)]
+pub unsafe extern "C" fn foxglove_mcap_attach(
+    writer: Option<&mut FoxgloveMcapWriter>,
+    attachment: &FoxgloveMcapAttachment,
+) -> FoxgloveError {
+    let Some(writer) = writer else {
+        tracing::error!("foxglove_mcap_attach called with null writer");
+        return FoxgloveError::ValueError;
+    };
+
+    match unsafe { do_foxglove_mcap_attach(writer, attachment) } {
+        Ok(()) => FoxgloveError::Ok,
+        Err(e) => e.into(),
+    }
+}
+
+unsafe fn do_foxglove_mcap_attach(
+    writer: &mut FoxgloveMcapWriter,
+    attachment: &FoxgloveMcapAttachment,
+) -> Result<(), foxglove::FoxgloveError> {
+    let name = unsafe { attachment.name.as_utf8_str() }.map_err(|e| {
+        foxglove::FoxgloveError::Utf8Error(format!("attachment name is invalid: {e}"))
+    })?;
+
+    let media_type = unsafe { attachment.media_type.as_utf8_str() }.map_err(|e| {
+        foxglove::FoxgloveError::Utf8Error(format!("attachment media_type is invalid: {e}"))
+    })?;
+
+    let Some(writer_handle) = writer.0.as_ref() else {
+        return Err(foxglove::FoxgloveError::SinkClosed);
+    };
+
+    let data = if attachment.data.is_null() || attachment.data_len == 0 {
+        &[]
+    } else {
+        unsafe { std::slice::from_raw_parts(attachment.data, attachment.data_len) }
+    };
```

Comment:
This is minor, but what if `attachment.data != NULL` and `attachment.data_len == 0`, or `attachment.data == NULL` and `attachment.data_len != 0` ?  In both cases, no data will be copied (so we'll avoid UB), but they are both nonsensical operations.  Should we explicitly return an error for these cases?